### PR TITLE
Implemented Mutator API

### PR DESCRIPTION
**Pull request by simondel**
I have finished implementing the Mutator API.

All `Mutation` classes are now `Mutator` classes.
The `Mutator` class is now called `MutatorOrchestrator`.

A Mutator now has to return a list of `ESTree.Node` objects, each of which has an ID (see: *src/api/mutant/StrykerNode.ts*). The `MutatorOrchestrator` now creates the `Mutants` instead of handing this task to the `Mutators`.

Mutators can return a **childnode** of the `Node` they received. This is already done in the `RemoveConditionalsMutator`.

I have also added a Mutator(see: *src/mutators/BlockStatementMutator.ts*) from grunt-mutation-testing which removes the entire content of a block-statement. This includes function definitions.

### CODE REVIEW

File: customTyping/escodegen.d.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
+
+declare module 'escodegen' {
```
Comment (nicojs) [inline customTyping/escodegen.d.ts]: Can you also make a PR for this in the DefinitelyTyped repo? That way we can remove this in the future.

File: src/MutatorOrchestrator.ts [hunk 2]
```diff
@@ -28,24 +29,22 @@ export default class Mutator {
...
+    let mutants: Mutant[] = [];
 
-    _.forEach(sourceFiles, (sourceFile: string) => {
+    sourceFiles.forEach((sourceFile: string) => {
       try {
-        var fileContent = fileUtils.readFile(sourceFile);
-        var abstractSyntaxTree = parserUtils.parse(fileContent);
-        var nodes = parserUtils.getNodesWithType(abstractSyntaxTree, types);
-        var newMutants = this.findMutants(sourceFile, fileContent, abstractSyntaxTree, nodes);
+        let fileContent = fileUtils.readFile(sourceFile);
```
Comment (nicojs) [inline src/MutatorOrchestrator.ts]: Can't we use an async file io call? Using blocking io can be time consuming when we have hunderds of source files. On the other hand, we can also wait for it to become a problem.

File: src/MutatorOrchestrator.ts [hunk 3]
```diff
@@ -28,24 +29,22 @@ export default class Mutator {
    * @returns {Mutant[]} The generated Mutants.
    */
   mutate(sourceFiles: string[]) {
```
Comment (nicojs) [inline src/MutatorOrchestrator.ts]: we should maybe rename this to `generateMutants`. I think that name more accurately describes what the method is doing.

File: src/MutatorFactory.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import {Mutator} from './api/mutant';
```
Comment (nicojs) [inline src/MutatorFactory.ts]: Cool :+1:

We should move this file in the api folder, that way we can later easily seperate it from the stryker main repo into the stryker-api repo.

File: src/utils/parserUtils.ts [hunk 5]
```diff
@@ -34,30 +33,41 @@ export function parse(code: string): any {
...
 /**
- * Finds all nodes which have one of several types in a syntax tree.
+ * Finds all nodes which a type in a syntax tree.
  * @function
  * @param abstractSyntaxTree - The current part of the abstract syntax tree which will be investigated.
- * @param  types - The list of types which are requested.
- * @returns  All nodes which have one of the requested types.
+ * @returns  All nodes with a type.
  */
-export function getNodesWithType(abstractSyntaxTree: any, types: string[], nodes?: any[], key?: string): any[] {
```
Comment (nicojs) [inline src/utils/parserUtils.ts]: This method name is not descriptive. What is it doing? Is the key-parameter ever used? Maybe it should be called `collectNodes` or something, or `collectAndFreezeNodes` (although i'm not a fan of `and` in method names). What do you think?

File: src/MutatorOrchestrator.ts [hunk 6]
```diff
@@ -27,25 +27,23 @@ export default class Mutator {
...
+        let nodes = parserUtils.getNodesWithType(abstractSyntaxTree);
+        let newMutants = this.findMutants(sourceFile, fileContent, abstractSyntaxTree, nodes);
         mutants = mutants.concat(newMutants);
       } catch (err) {
         switch (err.code) {
           case 'ENOENT':
             log.info(`Skipping file ${err.path} because it does not exist`);
             break;
           default:
+            console.log(err);
```
Comment (nicojs) [inline src/MutatorOrchestrator.ts]: We should use log.error for this (using a log4js logger)

File: src/MutatorOrchestrator.ts [hunk 7]
```diff
@@ -64,13 +62,21 @@ export default class Mutator {
...
-        _.forEach(this.mutations, (mutation: BaseMutation) => {
-          if (mutation.canMutate(astnode)) {
-            mutants = mutants.concat(mutation.applyMutation(sourceFile, originalCode, astnode, ast));
+        Object.freeze(astnode);
+        this.mutators.forEach((mutator: Mutator) => {
+          try {
+            mutator.applyMutations(astnode, _.cloneDeep).forEach((mutatedNode: ESTree.Node) => {
+              let mutatedCode = parserUtils.generate(mutatedNode);
+              let originalNode = nodes[(<StrykerNode>mutatedNode).nodeID];
+              mutants.push(new Mutant(mutator, sourceFile, originalCode, mutatedCode, originalNode.loc));
```
Comment (nicojs) [inline src/MutatorOrchestrator.ts]: Maybe add a debug log line here?
`generated 3 mutants on using source line 23 using the RemoveCondition mutator` for example?

File: src/Stryker.ts [hunk 8]
```diff
@@ -92,6 +93,17 @@ export default class Stryker {
     });
   }
 
+  private loadDefaultPlugins() {
```
Comment (nicojs) [inline src/Stryker.ts]: Maybe we should not call this 'default plugins'. Maybe we could better call it `loadDefaultMutators` and add it as a method to the `MutatorOrchestrator`. I don't see a single mutator as being a plugin. So also don't use the `PluginLoader`. That way, a mutators is no longer responsible by itself to add it to the `MutatorFactory` (that should generaly only be done by a plugin)

File: src/api/mutant/Mutator.ts [hunk 9]
```diff
@@ -14,9 +14,10 @@ interface Mutator {
    * If the mutator wants to mutate the node, it should return a clone of the node with mutations,
    * otherwise null. 
    * @param node A FROZEN Node which could be cloned and mutated.
+   * @param deepCopy A function to create a deep copy of an object.
    * @returns An array of mutated Nodes.
    */
-  applyMutations(node: ESTree.Node): ESTree.Node[];
+  applyMutations(node: ESTree.Node, deepCopy: (obj: any) => any): ESTree.Node[];
```
Comment (nicojs) [inline src/api/mutant/Mutator.ts]: As an aside: I think a shallow copy is enough. You generally only want to alter the current node anyway. The children can still point to the frozen nodes. Maybe we can add a copy function with a `boolean` parameter (`deep`).

File: src/api/mutant/StrykerNode.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+interface StrykerNode extends ESTree.Node {
+  nodeID: number;
```
Comment (nicojs) [inline src/api/mutant/StrykerNode.ts]: Ok, i might be nitpicking here, but i think the casing should be `nodeId` in this case. `Id` is short for `Identifier`. Abbreviations should always be written in PascalCase (in contrast to 2 letter acronyms). But i might be alone in this ;) so you can leave it if you want.

File: src/mutators/BlockStatementMutator.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import {Syntax} from 'esprima';
+import {Mutator, MutatorFactory} from '../api/mutant';
+
+/**
+ * Represents a mutator which can remove the content of a BlockStatement.
+ */
+export default class BlockStatementMutator implements Mutator {
+  name = 'BlockStatement';
+  types = [Syntax.BlockStatement];
```
Comment (nicojs) [inline src/mutators/BlockStatementMutator.ts]: by default, attributes are public. I think it should be private in this case

File: src/mutators/BlockStatementMutator.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+    return nodes;
+  }
+
+  private canMutate(node: ESTree.Node) {
+    return !!(node && this.types.indexOf(node.type) >= 0);
+  };
+}
+
+MutatorFactory.instance().register('BlockStatement', BlockStatementMutator);
```
Comment (nicojs) [inline src/mutators/BlockStatementMutator.ts]: Like stated before: I think a mutator should not behave as a plugin. The adding to the `MutatorFactory` might be a job for the MutatorOrchestrator?

Comment (simondel) [inline src/mutators/BlockStatementMutator.ts]: Would you like this to be the case for all *default* Mutators? (So all supported Mutators as of right now)

Comment (nicojs) [inline src/mutators/BlockStatementMutator.ts]: Yes i would. I think we can later support a black list to disable certain mutators.

File: customTyping/escodegen.d.ts (hunk 1 above)
Comment (simondel) [inline customTyping/escodegen.d.ts]: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/9326 :+1:

Comment (nicojs) [inline customTyping/escodegen.d.ts]: Cool! :+1: Usually they are pretty good with accepting new type definitions

File: src/api/mutant/StrykerNode.ts (hunk 10 above)
Comment (simondel) [inline src/api/mutant/StrykerNode.ts]: Well I honestly don't care if we use ID or Id, but I think we should be consistent. Shall I change it to `Id`?

File: src/utils/parserUtils.ts (hunk 5 above)
Comment (simondel) [inline src/utils/parserUtils.ts]: I think `collectNodes` is a good name. The comments can specify that nodes are also frozen. The `key` parameter was still left over from an earlier implementation of nodes, nice find!

File: src/Stryker.ts (hunk 8 above)
Comment (simondel) [inline src/Stryker.ts]: This method has become obsolete due to ca6cfd15603bc151619d4b51526f070adb773dc4

File: src/utils/parserUtils.ts (hunk 5 above)
Comment (nicojs) [inline src/utils/parserUtils.ts]: Maybe `collectFrozenNodes` is an even better name? :)

File: src/api/mutant/StrykerNode.ts (hunk 10 above)
Comment (nicojs) [inline src/api/mutant/StrykerNode.ts]: We'll leave it like it is, `nodeID`

**Comment by simondel** [comment]
@nicojs I believe that all feedback has now been implemented :+1:

**Pull request merged**